functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { tokens } = await oauth2Client.getToken(req.query.code);
+    oauth2Client.setCredentials(tokens);
+
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const profile = await gm.users.getProfile({ userId: 'me' });
+    const emailAddressKey = profile.data.emailAddress
+      .split('@')[0]
+      .replace('.', '');
+
+    await saveTokens({
+      emailKey: emailAddressKey,
+      fullEmail: profile.data.emailAddress,
+      accessToken: tokens.access_token,
+      refreshToken: tokens.refresh_token,
+    });
+
+    res.redirect(
+      `/Gmail-initWatch?emailAddress=${querystring.escape(profile.data.emailAddress)}`
+    );
+  }
+);
+
+const fetchToken = async (email: string) => {
+  const emailKey = email.split('@')[0].replace('.', '');
+  const queryResults = await db.ref(`/gmail/${emailKey}`).once('value');
+  return queryResults.val();
+};
+
+const storeHistoryIdInDatabase = async (email: string, historyId: any) => {
+  const emailKey = email.split('@')[0].replace('.', '');
+  return await db.ref(`/gmail/${emailKey}`).update({
+    lastSyncHistoryId: historyId,
+  });
+};
+
+export const initWatch = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    if (!req.query.emailAddress) {
+      return res.status(400).send('No emailAddress specified.');
+    }
+    const email = querystring.unescape(req.query.emailAddress);
+    if (!email.includes('@')) {
+      return res.status(400).send('Invalid emailAddress.');
+    }
+
+    // Initiate gmail client
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+
+    const { oauth, emailAddress } = await fetchToken(email);
+
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    try {
+      // Make sure we have the correct labelId because GMail internally gives a random one that is
+      // not based on the given name, so we need to filter it by the given name
+      const labelResults = await gm.users.labels.list({
+        userId: emailAddress,
+      });
+      const doneLabelObj = labelResults.data.labels.filter(label => {
+        return label.name === 'SQRDone';
+      })[0];
+
+      const watchResults = await gm.users.watch({
+        userId: 'me',
+        requestBody: {
+          labelIds: [doneLabelObj.id],
+          topicName: 'projects/audio-seva-team-test/topics/gmail-labeled-done',
+        },
+      });
+
+      if (watchResults.data && doneLabelObj) {
```

Comment:
Can we inverse the condition and do a quick exit here? I don't like such `if`s.

Reply:
Not fixed.

Reply:
Yes, that makes sense. Looks like this now:

```
      if (!watchResults.data || !doneLabelObj) {
        return res
          .status(301)
          .send(`Something went wrong, check logs for Gmail-initWatch`);
      }

      // Store the most recent historyId
      await storeHistoryIdInDatabase(watchResults.data.historyId);
      return res
        .status(200)
        .send(
          `Watch (${watchResults.data.historyId}) created on ${
            doneLabelObj.name
          }/${doneLabelObj.id} for coordinator`
        );
```